### PR TITLE
Bugfixes for output generation in case of series failure

### PR DESCRIPTION
**Pull request by govarsha**
Made changes so that report gets generated even if forecast fails for any series, any model
For arima, when trend is constant forecast was failing if there are constant columns and for any trend if there are boolean columns it fails. So added bool in label_encoder so that boolean can be converted and added constant columns removal for constant trend in arima.
Explanations were failing when data had boolean columns (fixed it by adding label_encoder)
`TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced 
     any supported types according to the casting rule ''safe''`

Added unit tests to check above errors.

### CODE REVIEW

File: ads/opctl/operator/lowcode/forecast/model/arima.py [hunk 1]
```diff
@@ -64,6 +65,10 @@ def _train_model(self, i, s_id, df, model_kwargs):
         try:
             target = self.original_target_column
             self.forecast_output.init_series_output(series_id=s_id, data_at_series=df)
+            # If trend is constant, remove constant columns
+            if 'trend' not in model_kwargs or model_kwargs['trend'] == 'c':
+                self.constant_cols[s_id] = df.columns[df.nunique() == 1]
+                df = df.drop(columns=self.constant_cols[s_id])
```
Comment (ahosler) [inline ads/opctl/operator/lowcode/forecast/model/arima.py:71]: This is very nice and succint 👍

File: ads/opctl/operator/lowcode/forecast/model/base_model.py [hunk 2]
```diff
@@ -650,38 +654,49 @@ def explain_model(self):
...
-
-            kernel_explnr = PermutationExplainer(
-                model=explain_predict_fn, masker=data_trimmed
-            )
-            kernel_explnr_vals = kernel_explnr.shap_values(data_trimmed)
-
-            exp_end_time = time.time()
-            global_ex_time = global_ex_time + exp_end_time - exp_start_time
+                explain_predict_fn = self.get_explain_predict_fn(series_id=s_id)
+                if self.spec.model == SupportedModels.Arima and s_id in self.constant_cols:
```
Comment (ahosler) [inline ads/opctl/operator/lowcode/forecast/model/base_model.py:660]: Can we move both of these Arima logic sections into the arima file under the explanations section?

File: ads/opctl/operator/lowcode/forecast/model/base_model.py [hunk 3]
```diff
@@ -700,9 +715,17 @@ def local_explainer(self, kernel_explainer, series_id, datetime_col_name) -> Non
...
         data = self.datasets.get_horizon_at_series(s_id=series_id)
-
+        # columns that were dropped in train_model in arima, should be dropped here as well
+        if self.spec.model == SupportedModels.Arima and series_id in self.constant_cols:
+            data = data.drop(columns=self.constant_cols[series_id])
         data[datetime_col_name] = datetime_to_seconds(data[datetime_col_name])
         data = data.reset_index(drop=True)
+
+        # Explainer fails when boolean columns are passed for arima
+        if self.spec.model == SupportedModels.Arima:
```
Comment (ahosler) [inline ads/opctl/operator/lowcode/forecast/model/base_model.py:725]: These 2 as well? Or is there a reason we can't?

File: tests/operators/forecast/test_errors.py [hunk 4]
```diff
@@ -476,5 +497,148 @@ def split_df(df):
...
+        assert os.path.exists(report_path), f"Report file not found at {report_path}"
+
+        error_path = f"{output_data_path}/errors.json"
+        assert os.path.exists(error_path), f"Error file not found at {error_path}"
+
+        # Additionally, you can read the content of the error.json and assert its content
+        with open(error_path, 'r') as error_file:
+            error_content = json.load(error_file)
+            assert "Custom exception message" in error_content["1"]["error"], "Error message mismatch"
+            assert "Custom exception message" in error_content["13"]["error"], "Error message mismatch"
```
Comment (ahosler) [inline tests/operators/forecast/test_errors.py:545]: This is a great addition

**Review by ahosler — DISMISSED** [review]
great work! If we can refactor arima that would be awesome, but if we can't then this is ready to merge!

File: ads/opctl/operator/lowcode/forecast/model/base_model.py (hunk 2 above)
Comment (govarsha) [reply ads/opctl/operator/lowcode/forecast/model/base_model.py]: Yeah, moved it

File: ads/opctl/operator/lowcode/forecast/model/base_model.py (hunk 3 above)
Comment (govarsha) [reply ads/opctl/operator/lowcode/forecast/model/base_model.py]: Moved it

**Review by ahosler — APPROVED** [review]

**Review by ahosler — APPROVED** [review]

**Review by mayoor — APPROVED** [review]

**Pull request merged**